packages/vaadin-combo-box/src/vaadin-combo-box-dropdown.js

```diff
@@ -35,6 +35,17 @@ class ComboBoxOverlayElement extends OverlayElement {
     return 'vaadin-combo-box-overlay';
```

Comment:
The issue still seems to be there. Also on `master`:

https://user-images.githubusercontent.com/1222264/143578418-e7fceffe-8fe2-4ad3-af40-34669feabf4e.mp4

Reply:
```html
<!DOCTYPE html>
<html lang="en">
  <head>
    <meta charset="UTF-8" />
    <meta http-equiv="X-UA-Compatible" content="IE=edge" />
    <meta name="viewport" content="width=device-width, initial-scale=1.0" />
    <title>Document</title>
    <script type="module">
      import '@vaadin/vaadin-combo-box/vaadin-combo-box.js';
      import '@vaadin/vaadin-dialog/vaadin-dialog.js';

      const dialog = document.querySelector('vaadin-dialog');
      dialog.renderer = function(root, dialog) {
        root.innerHTML = `
          <vaadin-combo-box items='[1,2,3]'></vaadin-combo-box>
        `;
      };
    </script>
  </head>
  <body>
    <vaadin-dialog opened></vaadin-dialog>
  </body>
</html>
```

Reply:
Ah, it's about clicking _inside_ the dialog

Reply:
Interesting - I was testing with the time picker, because that's the component that the issue was reported for. It's definitely reproducable there:
```html
<!DOCTYPE html>
<html lang="en">
<head>
  <meta charset="UTF-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <meta name="viewport" content="width=device-width, initial-scale=1.0" />
  <title>Dialog</title>
</head>
<body>
<vaadin-button>Open Dialog</vaadin-button>
<vaadin-dialog>
</vaadin-dialog>
<script type="module">
  import "@vaadin/vaadin-button";
  import "@vaadin/vaadin-dialog";
  import "@vaadin/vaadin-time-picker";

  const button = document.querySelector("vaadin-button");
  const dialog = document.querySelector("vaadin-dialog");
  dialog.renderer = (root) => {
    root.innerHTML = "<vaadin-time-picker></vaadin-time-picker>";
    root.querySelector('vaadin-time-picker');
  }
  button.addEventListener("click", () => dialog.opened = true);
</script>
</body>
</html>
```

Reply:
As for the dialog closing when clicking on the backdrop: That could be interpreted as expected I guess, and this issue was about clicking into the dialog. I would say that the dialog probably shouldn't close in that case as well, but that would require further effort. 

One thing to note here is that clicking on the backdrop only closes the dialog, if the combobox / time picker input field is focused. For combo box that happens always when opening the dropdown, even when clicking on the toggle button. For time picker that does not happen when clicking on the toggle button - that is probably a different issue.

Reply:
There's a [new test case](https://github.com/vaadin/web-components/pull/3110/files#diff-a2eb130bdaf784902d00661e9cfd123978e395cfbdf1f35b4457c806a92d61e2R63-R64) that seems to be verifying that the dialog wouldn't close on backdrop click however.

Reply:
Interestingly, it seems to behave differently, depending on if you open the dropdown by the input or by the icon

https://user-images.githubusercontent.com/1222264/143583231-7e2852f3-964b-40c2-9841-4c1c7ba717b0.mp4

Reply:
Oh dear...I'm going to investigate this some more on Monday.

Reply:
@tomivirkki 

Did some debugging on the flow of events. It's a bit tricky, I hope I can convey what's going on.

Some things to note beforehand how the OverlayElement works:

- Opening a modal overlay disables pointer-events on lower-level overlays, and restores them when they are closed
- The outside click listener in OverlayElement expects that it will be run from lowest-level overlay to highest-level overlay. See [this condition](https://github.com/vaadin/web-components/blob/21.0/packages/vaadin-overlay/src/vaadin-overlay.js#L502) in OverlayElement. This usually works, because the listeners are registered when opening an overlay. If the listeners were processed from highest-level to lowest-level, then `last` would always be true, and all overlays would be closed by a single click.

**Why does the time picker not close the dialog when the input is focused?**

Clicking outside will trigger the focus out event first, which already closes the time picker dropdown, and restores pointer-events on the dialog overlay. Due to restoring pointer-events, the browser registers a mouseup event on the dialog content (https://github.com/vaadin/web-components/blob/21.0/packages/vaadin-overlay/src/vaadin-overlay.js#L483). The mouseup event causes the dialog to not close (https://github.com/vaadin/web-components/blob/21.0/packages/vaadin-overlay/src/vaadin-overlay.js#L497).

**Why does the time picker close the dialog when the input is not focused, for example when clicking the toggle button?**

Time picker uses combo box, which has a custom outside click listener, **which runs in capture phase**. That means that it's custom outside click listener will run before the one for OverlayElement, which breaks the precondition that the listeners must be run from lowest-level to highest-level. When the time picker overlay closes, it makes the dialog overlay the last one. The outside click listener for the dialog overlay runs afterwards. It determines that it's the last overlay, and that the click was not inside the dialog, because the click event was registered at a time when the pointer-events have not been restored. Thus it closes.

**Why does it work for combo box itself then?**

Because combo box always focuses the input when opening the dropdown, even when clicking the toggle button (this is different from time picker, which does not focus the input in this case). That means it's always closed by the focus out event.

**Summary**

The main issue seems to be the custom outside click listener of combo box running in capture phase. Removing it, and reusing the outside click listener of OverlayElement fixes the issue, because the listeners are now run in the expected order.

Focus out is still and issue on its own, in that closing the combo box dropdown can also close the dialog, when clicking outside of it. IMO a single click should only close a single overlay, always. But it seems tricky to solve, and it's also not that important compared the issue that is fixed here.